sw/host/opentitanlib/src/transport/dediprog/spi.rs

```diff
@@ -0,0 +1,367 @@
+                "Command not supported".to_string()
+            )),
+        }
+        usbcmd[6..10].clone_from_slice(&cmd.get_address().to_le_bytes());
+        self.inner.borrow().device.write_control(
+            rusb::request_type(
+                rusb::Direction::Out,
+                rusb::RequestType::Vendor,
+                rusb::Recipient::Endpoint,
+            ),
+            Command::Write as u8,
+            0,
+            0,
+            &usbcmd,
+        )?;
+        for chunk in wbuf.chunks_exact(Self::WRITE_CHUNK_SIZE) {
+            let mut buf = [0xFFu8; 512];
+            buf[0..Self::WRITE_CHUNK_SIZE].clone_from_slice(&chunk);
+            self.inner
+                .borrow()
+                .device
+                .write_bulk(self.inner.borrow().out_endpoint, &buf)?;
+        }
+        Ok(())
+    }
+
+    fn do_run_eeprom_transactions(
+        &self,
+        mut transactions: &mut [eeprom::Transaction],
+    ) -> Result<()> {
+        loop {
+            match transactions {
+                [eeprom::Transaction::Command(pre_cmd), Write(cmd, ref wbuf), WaitForBusyClear, rest @ ..] =>
+                {
+                    transactions = rest;
+                    if pre_cmd.get_opcode() == [0x06] {
+                        // Write enable is done by eeprom_write_transaction()
+                    } else {
+                        self.run_transaction(&mut [Transfer::Write(cmd.to_bytes()?)])?
+                    }
+                    self.eeprom_write_transaction(cmd, wbuf, true)?;
+                }
+                [eeprom::Transaction::Command(cmd), rest @ ..] => {
+                    transactions = rest;
+                    self.run_transaction(&mut [Transfer::Write(cmd.to_bytes()?)])?
+                }
+                [Read(cmd, ref mut rbuf), rest @ ..] => {
+                    transactions = rest;
+                    self.eeprom_read_transaction(cmd, rbuf)?;
+                }
+                [Write(cmd, ref wbuf), WaitForBusyClear, rest @ ..] => {
+                    transactions = rest;
+                    self.eeprom_write_transaction(cmd, wbuf, true)?;
+                }
+                [Write(cmd, ref wbuf), rest @ ..] => {
+                    transactions = rest;
+                    self.eeprom_write_transaction(cmd, wbuf, false)?;
+                }
+                [WaitForBusyClear, rest @ ..] => {
+                    transactions = rest;
+                    while {
+                        let mut buf = [0u8; 1];
+                        self.run_transaction(&mut [
+                            Transfer::Write(&[eeprom::READ_STATUS]),
+                            Transfer::Read(&mut buf),
+                        ])?;
+                        buf[0]
+                    } & eeprom::STATUS_WIP
+                        != 0
+                    {}
+                }
+                [] => return Ok(()),
+            }
+        }
+    }
+}
+
+impl Target for DediprogSpi {
```

Comment:
The `Target` trait has a `set_voltage` function.  We should add an implementation so the voltage options specified with `io::spi::SpiParams` get applied at init time.

Reply:
I had failed to recognize that trait function, of course it should be implemented. Done.

(I should probably add voltage and spi mode fields to the json format, as well.  That can come later.)